tcptracer-bpf.c

```diff
@@ -864,17 +853,94 @@ int kretprobe__inet_csk_accept(struct pt_regs *ctx)
 		return 0;
 	}
 
+    struct fd_info t = { .active = 1 };
+    t.start_time_ns = bpf_ktime_get_ns();
+
+    bpf_map_update_elem(&active_fds, &newsk, &t, BPF_ANY);
+
 	return assert_tcp_record(newsk, status, DIRECTION_INCOMING, STATE_INITIALIZING);
 }
 
+// Is kretprobe__inet_csk_accept sufficient or is this one needed as well or instead?
+
+// SEC("kprobe/__x64_sys_accept4")
+// int kprobe__x64_sys_accept4(struct pt_regs *ctx)
+// {
+	// The file descriptor is the value returned from the syscall.
+	// int fd = PT_REGS_RC(ctx);
+	// bool t = true;
+	// bpf_map_update_elem(&active_fds, &fd, &t, BPF_ANY);
+
+// 	return 0;
+// }
+
+SEC("kprobe/__x64_sys_writev")
+int kprobe__sys_writev(struct pt_regs *ctx)
+{
+    int fd = PT_REGS_PARM1(ctx);
+    struct iovec *vectors = (struct iovec *)PT_REGS_PARM2(ctx);
+
+    struct fd_info *res = bpf_map_lookup_elem(&active_fds, &fd);
+    if (res == NULL) {
+//		 This file descriptor is not for a socket we're tracking
+        return 0;
+    }
+
+    int zero = 0;
+    void *data = bpf_map_lookup_elem(&write_buffer_heap, &zero);
+    if (data == NULL) {
+        return 0;
+    }
+
+    struct iovec vec = {};
+    bpf_probe_read(&vec, sizeof(struct iovec), vectors);
+	return 0;
+}
+
+
 SEC("kprobe/tcp_sendmsg")
 int kprobe__tcp_sendmsg(struct pt_regs *ctx) {
+
 	struct sock *sk = (struct sock *) PT_REGS_PARM1(ctx);
-	size_t size = (size_t) PT_REGS_PARM3(ctx);
+	struct msghdr *k_msg = (void *) PT_REGS_PARM2(ctx);
+	const size_t size = (size_t) PT_REGS_PARM3(ctx);
 	u64 zero = 0;
 
-	// TODO: Add DEBUG macro so this is only printed, if enabled
-	// bpf_debug("map: tcp_send_ipv4 kprobe\n");
+
+    struct fd_info *res = bpf_map_lookup_elem(&active_fds, &sk);
+    if (res == NULL) {
+//		 This file descriptor is not for a socket we're tracking
+        return 0;
+    }
+
+    char *data = bpf_map_lookup_elem(&write_buffer_heap, &zero);
+    if (data == NULL) {
+        return 0;
+    }
+
+    const char http_marker[4] = "HTTP";
+
+    u64 ttfb = bpf_ktime_get_ns() - res->start_time_ns;
+
+    struct msghdr msg = {};
+    bpf_probe_read(&msg, sizeof(msg), k_msg);
+    if ((msg.msg_iter.type & 4) == 4) {
+        struct iovec iov = {};
+        bpf_probe_read(&iov, sizeof(iov), (void *) msg.msg_iter.iov);
+        bpf_probe_read(data, MAX_MSG_SIZE, iov.iov_base);
+
+        int status_code = 100*(data[9]-'0') + 10*(data[10]-'0') + (data[11]-'0');
```

Comment:
Shouldn't we only grab the status code after asserting that this is an HTTP response?

Reply:
We only check "HTTP" prefix of a packet and then get the http code.